### PR TITLE
Chunk messages when enqueuing into Redis

### PR DESCRIPTION
**Pull request by Sneagan**
### Summary

fixes https://github.com/brave-intl/bat-go/issues/2218

### Type of Change

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: libs/redisconsumer/consumer_test.go
Comment (pavelbrm) [inline libs/redisconsumer/consumer_test.go:8]: It's good to have a test. However, I can suggest two improvements:
1. Do not reuse the `testMessages` variable. That's a well-known thing. And while reusing it here does not cause any meaningful harm, it does not bring any good at the same time. So let's don't do that. Unfortunately, this code base has got many of such examples, and adding more does not look good to me.
2. We've got multiple test cases here, but everything is coupled together under the single test. I strongly suggest, at least, using `t.Run` for each scenario. Supposedly, there should be a few test cases:
    - 0 elements
    - under the limit elements;
    - over the limit elements.

File: libs/redisconsumer/consumer.go
Comment (pavelbrm) [inline libs/redisconsumer/consumer.go:151]: This can be slightly tidied up and micro-, if you will, optimised. The `messages` slice is passed in, so it's likely that calling `len` multiple times on it will [incur overhead](https://stackoverflow.com/questions/26634554/go-multiple-len-calls-vs-performance).

Additionally, the size of a portion is better passed in as a parameter, so the code is easier to adjust at the call site.

1. If it were me, I would have written it as follows, also improving on readability:

```suggestion
func chunkMessages(lim int, messages []interface{}) [][]interface{} {
	total := len(messages)
	if total == 0 {
		return nil
	}

	result := make([][]interface{}, 0, 1)
	for start, stop := 0, 0; start < total; start = stop {
		stop += lim
		if stop > total {
			stop = total
		}

		result = append(result, messages[start:stop])
	}

	return result
}
```

2. Or even in generic form:

```suggestion
func splitSlice[T any](lim int, slice []T) [][]T {
	total := len(slice)
	if total == 0 {
		return nil
	}

	result := make([][]T, 0, 1)
	for start, stop := 0, 0; start < total; start = stop {
		stop += lim
		if stop > total {
			stop = total
		}

		result = append(result, slice[start:stop])
	}

	return result
}
```

3. Your current version could be slightly changed like this:

```suggestion
func chunkMessages(messages []interface{}) [][]interface{} {
	const chunkSize = 10000

	total := len(messages)
	var chunks [][]interface{}

	for i := 0; i < total; i += chunkSize {
		end := i + chunkSize
		if total < end {
			end = total
		}

		chunks = append(chunks, messages[i:end])
	}

	return chunks
}
```

File: libs/redisconsumer/consumer.go
Comment (pavelbrm) [inline libs/redisconsumer/consumer.go:119]: This `err` here does not make sense anymore. Now errors can occur, are checked and returned from inside the loop. So you can remove this from here, and also replace the final line with `return nil`.

**Review by pavelbrm — COMMENTED** [review]
Left some feedback. The current implementation is looking OK, but could be improved before merging. 🙂

File: libs/redisconsumer/consumer.go
Comment (Sneagan) [reply libs/redisconsumer/consumer.go]: Good recommendation

File: libs/redisconsumer/consumer_test.go
Comment (pavelbrm) [inline libs/redisconsumer/consumer_test.go:33]: This should do just as well:

```suggestion
	must.Equal(t, testMessages, testChunks[0])
```

File: libs/redisconsumer/consumer_test.go
Comment (pavelbrm) [inline libs/redisconsumer/consumer_test.go:57]: Likewise, here we can simplify:

```suggestion
	must.Equal(t, testMessages[0:500], testChunks[0])
	must.Equal(t, testMessages[500:], testChunks[1])
```

File: libs/redisconsumer/consumer_test.go
Comment (pavelbrm) [inline libs/redisconsumer/consumer_test.go:88]: I hope you get the idea. 😸

File: tools/payments/client.go
Comment (pavelbrm) [inline tools/payments/client.go:152]: Non-blocking nit: Here and below, the `to redis` thing seems like an unnecessary detail.

**Review by pavelbrm — APPROVED** [review]
Approving with non-blocking comments.

**Pull request merged**
